scraper/scraper_bot.py

```diff
@@ -172,130 +166,249 @@ def schema(self):
+            repo_type="dataset",
             commit_message="Rename chunks",
-            token=os.environ['HF_TOKEN'],
+            token=os.environ["HF_TOKEN"],
             operations=operations,
         )
-        # split into multiple commits. We can't split the copy and delete operations into separate commits
 
-    def _load_dataset(self) -> (Dataset, int):
+    def _load_dataset(self, schema: dict) -> (Dataset, int):
         chunks = self._get_chunk_names()
         if len(chunks) == 0:
+            # Initialize the dataset if empty or non-existent
+            self._create_repo()
             return (None, 0)
-        # sort in descending order. The naming scheme is  train-<x>-of-<y>-<hash>.parquet
+
+        # sort in descending order.
+        # The naming scheme is  train-<x>-of-<y>-<hash>.parquet
         chunks.sort(key=lambda x: int(x.split("-")[1]), reverse=True)
-        df = pd.read_parquet(fs.open(f"{self.fs_path}/{chunks[0]}", "rb"))
-        chunk_num = int(chunks[0].split("-")[1])
+        chunk_num = len(chunks)  # y in the naming scheme
+        print(f"Found {len(chunks)} chunks: {chunks}")
 
-        # This is a temporary fix for the transition period between the old and new saving schemes
-        # Search all chunks for the max message id
-        # TODO: Remove this once we are sure that all datasets have a new chunk - that guarantees that the latest message id is in the last chunk
-        latest_message_id = '0'
-        for chunk in chunks:
-            _df = pd.read_parquet(fs.open(f"{self.fs_path}/{chunk}", "rb"))
-            _latest_message_id = get_latest_message_id(df)
-            if _latest_message_id and _latest_message_id > latest_message_id:
-                latest_message_id = _latest_message_id
-                df = _df
-                chunk_num = int(chunk.split("-")[1])
+        print(f"Loading and converting to Hugging Face Dataset...")
+        ds = load_dataset(
+            self.hf_dataset_name, columns=schema, split="train", streaming=True
+        )
+        df = pd.DataFrame(ds)
 
         return (df, chunk_num)
 
-    def filter_messages(self, messages: List[HFDatasetScheme]) -> List[HFDatasetScheme]:
+    def filter_messages(
+        self, dataset: pd.DataFrame, messages: List[HFDatasetScheme]
+    ) -> List[HFDatasetScheme]:
+        fs = HfFileSystem(token=os.environ["HF_TOKEN"], skip_instance_cache=True)
+
         # Iterate over the whole dataset and remove all messages that are already in the dataset
         chunks = self._get_chunk_names()
         if len(chunks) == 0:
             return messages
 
-        for chunk in tqdm(chunks, desc="Filtering messages", unit=" chunks"):
-            df = pd.read_parquet(fs.open(f"{self.fs_path}/{chunk}", "rb"))
-            existing_message_ids = df['message_id'].tolist()
-            messages = [msg for msg in messages if msg.message_id not in existing_message_ids]
+        existing_message_ids = dataset["message_id"].tolist()
```

Comment:
I added this chunk filtering because we can't guarantee in which chunk the latest messages will be. If we use the custom chunking, then we can guarantee that.

Reply:
I was having trouble getting it to work properly, my thinking with this is that we'll know we're not adding duplicates, and simply adding the new data to the incomplete chunk.